presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotColumnUtils.java

```diff
@@ -39,14 +51,60 @@ private PinotColumnUtils()
+                .map(columnName -> new PinotColumn(columnName, getPrestoTypeFromPinotType(pinotTableSchema.getFieldSpecFor(columnName)), isNullableColumnFromPinotType(pinotTableSchema.getFieldSpecFor(columnName)),
+                        getCommentFromPinotType(pinotTableSchema.getFieldSpecFor(columnName))))
                 .collect(toImmutableList());
     }
 
+    private static String getCommentFromPinotType(FieldSpec field)
+    {
+        return field.getFieldType().name();
+    }
+
+    private static boolean isNullableColumnFromPinotType(FieldSpec field)
+    {
+        return DEFAULT_PINOT_COLUMN_IS_NULLABLE;
+    }
+
     public static Type getPrestoTypeFromPinotType(FieldSpec field)
     {
         if (field.isSingleValueField()) {
-            return getPrestoTypeFromPinotType(field.getDataType());
+            switch (field.getFieldType()) {
+                case TIME:
+                    TimeFieldSpec timeFieldSpec = (TimeFieldSpec) field;
+                    TimeGranularitySpec outSpec = timeFieldSpec.getOutgoingGranularitySpec();
+                    if (outSpec != null) {
+                        if (outSpec.getTimeFormat().equalsIgnoreCase(TimeGranularitySpec.TimeFormat.EPOCH.name())) {
+                            if ((TimeUnit.DAYS == outSpec.getTimeType()) && (outSpec.getTimeUnitSize() == 1)) {
+                                return DateType.DATE;
+                            }
+                            if ((TimeUnit.MILLISECONDS == outSpec.getTimeType()) && (outSpec.getTimeUnitSize() == 1)) {
+                                return TimestampType.TIMESTAMP;
+                            }
+                        }
+                    }
+                    else {
+                        TimeGranularitySpec inSpec = timeFieldSpec.getIncomingGranularitySpec();
+                        if ((TimeUnit.DAYS == inSpec.getTimeType()) && (inSpec.getTimeUnitSize() == 1)) {
+                            return DateType.DATE;
+                        }
+                        if ((TimeUnit.MILLISECONDS == inSpec.getTimeType()) && (inSpec.getTimeUnitSize() == 1)) {
+                            return TimestampType.TIMESTAMP;
+                        }
+                    }
+                    return getPrestoTypeFromPinotType(field.getDataType());
+                case DATE_TIME:
+                    DateTimeFieldSpec dateTimeFieldSpec = (DateTimeFieldSpec) field;
+                    if (dateTimeFieldSpec.getFormat().equalsIgnoreCase(DAYS_SINCE_EPOCH_TIME_FORMAT)) {
+                        return DateType.DATE;
+                    }
+                    if (dateTimeFieldSpec.getFormat().equalsIgnoreCase(MILLISECONDS_SINCE_EPOCH_TIME_FORMAT)) {
+                        return TimestampType.TIMESTAMP;
+                    }
+                case DIMENSION:
+                case METRIC:
+                    return getPrestoTypeFromPinotType(field.getDataType());
```

Comment:
Maybe this can go outside the switch block ?

Reply:
sure